sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryAvroUtils.java

```diff
@@ -402,23 +403,55 @@ private static Object convertRequiredField(
         unionTypes.get(1).getType(), unionTypes.get(1).getLogicalType(), fieldSchema, v);
   }
 
-  static Schema toGenericAvroSchema(String schemaName, List<TableFieldSchema> fieldSchemas) {
+  static Schema toGenericAvroSchema(
+      String schemaName, List<TableFieldSchema> fieldSchemas, @Nullable String namespace) {
+
+    String nextNamespace = namespace == null ? null : String.format("%s.%s", namespace, schemaName);
+
     List<Field> avroFields = new ArrayList<>();
     for (TableFieldSchema bigQueryField : fieldSchemas) {
-      avroFields.add(convertField(bigQueryField));
+      avroFields.add(convertField(bigQueryField, nextNamespace));
     }
     return Schema.createRecord(
         schemaName,
         "Translated Avro Schema for " + schemaName,
-        "org.apache.beam.sdk.io.gcp.bigquery",
+        namespace == null ? "org.apache.beam.sdk.io.gcp.bigquery" : namespace,
         false,
         avroFields);
   }
 
+  static Schema toGenericAvroSchema(String schemaName, List<TableFieldSchema> fieldSchemas) {
+    return toGenericAvroSchema(
+        schemaName,
+        fieldSchemas,
+        hasNamespaceCollision(fieldSchemas) ? "org.apache.beam.sdk.io.gcp.bigquery" : null);
+  }
+
+  // To maintain backwards compatibility we only disambiguate collisions in the field namespaces as
+  // these never worked with this piece of code.
+  private static boolean hasNamespaceCollision(List<TableFieldSchema> fieldSchemas) {
```

Comment:
nit: this seems easy to unit test

Reply:
it really only exists for reading clarity (single call-site) so not sure what value another test would bring in addition to the test that already explicitly exercises the functionality.

Reply:
Ack -- it was a nit, as if there is a regression it will be easier to get to the bottom of a failure, but I agree with the single call-site argument / coverage by the other test -- it is relatively simple + static enough too.